File: ts/project/.sdk/tm/java/utility/MakeRequest.java
package JAVAPACKAGE.utility;

import java.util.LinkedHashMap;
import java.util.Map;

import JAVAPACKAGE.core.Context;
import JAVAPACKAGE.core.Response;
import JAVAPACKAGE.core.Result;
import JAVAPACKAGE.core.Spec;
import JAVAPACKAGE.core.Utility;

@SuppressWarnings({"unchecked"})
final class MakeRequest {

  private MakeRequest() {}

  static Response makeRequest(Context ctx) {
    Object outRequest = ctx.out.get("request");
    if (outRequest instanceof Response) {
      return (Response) outRequest;
    }

    Spec spec = ctx.spec;
    Utility utility = ctx.utility;

    Response response = new Response(new LinkedHashMap<>());
    Result result = new Result(new LinkedHashMap<>());
    ctx.result = result;

    if (spec == null) {
      throw ctx.makeError("request_no_spec",
          "Expected context spec property to be defined.");
    }

    Map<String, Object> fetchdef;
    try {
      fetchdef = utility.makeFetchDef.apply(ctx);
    }
    catch (RuntimeException err) {
      response.err = err;
      ctx.response = response;
      spec.step = "postrequest";
      return response;
    }

    if (ctx.ctrl.explain != null) {
      ctx.ctrl.explain.put("fetchdef", fetchdef);
    }

    spec.step = "prerequest";

    Object url = fetchdef.get("url");
    Object fetched = null;
    RuntimeException fetchErr = null;
    try {
      fetched = utility.fetcher.fetch(ctx,
          url instanceof String ? (String) url : "", fetchdef);
    }
    catch (RuntimeException err) {
      fetchErr = err;
    }

    if (fetchErr != null) {
      response.err = fetchErr;
    }
    else if (fetched == null) {
      Map<String, Object> resmap = new LinkedHashMap<>();
      resmap.put("err", ctx.makeError("request_no_response", "response: undefined"));
      response = new Response(resmap);
    }
    else if (fetched instanceof Map) {
      response = new Response((Map<String, Object>) fetched);
    }
    else {
      response.err = ctx.makeError("request_invalid_response", "response: invalid type");
    }

    spec.step = "postrequest";
    ctx.response = response;

    return response;
  }
}

File: ts/project/.sdk/tm/csharp/utility/MakeRequest.cs
// ProjectName SDK utility: makeRequest - build the fetch definition and
// call the transport. Transport failures are carried on the returned
// Response (Err) rather than thrown, mirroring the go pipeline.

namespace ProjectNameSdk.Util;

public static partial class SdkUtility
{
    internal static Response MakeRequestUtil(Context ctx)
    {
        if (ctx.Out.TryGetValue("request", out var outReq) && outReq is Response cached)
        {
            return cached;
        }

        var spec = ctx.Spec;
        var utility = ctx.Utility!;

        var response = new Response(new Dictionary<string, object?>());
        var result = new Result(new Dictionary<string, object?>());
        ctx.Result = result;

        if (spec == null)
        {
            throw ctx.MakeError("request_no_spec",
                "Expected context spec property to be defined.");
        }

        Dictionary<string, object?> fetchdef;
        try
        {
            fetchdef = utility.MakeFetchDef(ctx);
        }
        catch (Exception err)
        {
            response.Err = err;
            ctx.Response = response;
            spec.Step = "postrequest";
            return response;
        }

        if (ctx.Ctrl.Explain != null)
        {
            ctx.Ctrl.Explain["fetchdef"] = fetchdef;
        }

        spec.Step = "prerequest";

        var url = fetchdef.TryGetValue("url", out var u) ? u as string ?? "" : "";

        object? fetched = null;
        Exception? fetchErr = null;
        try
        {
            fetched = utility.Fetcher(ctx, url, fetchdef);
        }
        catch (Exception ex)
        {
            fetchErr = ex;
        }

        if (fetchErr != null)
        {
            response.Err = fetchErr;
        }
        else if (fetched == null)
        {
            response = new Response(new Dictionary<string, object?>
            {
                ["err"] = ctx.MakeError("request_no_response", "response: undefined"),
            });
        }
        else if (fetched is Dictionary<string, object?> fm)
        {
            response = new Response(fm);
        }
        else
        {
            response.Err = ctx.MakeError("request_invalid_response", "response: invalid type");
        }

        spec.Step = "postrequest";
        ctx.Response = response;

        return response;
    }
}
